src/View/SSViewer_Scope.php

```diff
@@ -45,107 +44,107 @@ class SSViewer_Scope
-     *
-     * @var Iterator
      */
-    protected $itemIterator;
+    protected ?Iterator $itemIterator;
 
     /**
      * Total number of items in the iterator
-     *
-     * @var int
      */
-    protected $itemIteratorTotal;
+    protected int $itemIteratorTotal;
 
     /**
      * A pointer into the item stack for the item that will become the active scope on the next pop call
-     *
-     * @var int
      */
-    private $popIndex;
+    private ?int $popIndex;
 
     /**
      * A pointer into the item stack for which item is "up" from this one
-     *
-     * @var int
      */
-    private $upIndex;
+    private ?int $upIndex;
 
     /**
      * A pointer into the item stack for which the active item (or null if not in stack yet)
-     *
-     * @var int
      */
-    private $currentIndex;
+    private int $currentIndex;
 
     /**
      * A store of copies of the main item stack, so it's preserved during a lookup from local scope
      * (which may push/pop items to/from the main item stack)
-     *
-     * @var array
      */
-    private $localStack = [];
+    private array $localStack = [];
 
     /**
      * The index of the current item in the main item stack, so we know where to restore the scope
      * stored in $localStack.
+     */
+    private int $localIndex = 0;
+
+    /**
+     * List of global property providers
      *
-     * @var int
+     * @internal
+     * @var TemplateGlobalProvider[]|null
      */
-    private $localIndex = 0;
+    private static $globalProperties = null;
 
     /**
-     * @var object $item
-     * @var SSViewer_Scope $inheritedScope
+     * List of global iterator providers
+     *
+     * @internal
+     * @var TemplateIteratorProvider[]|null
      */
-    public function __construct($item, SSViewer_Scope $inheritedScope = null)
-    {
+    private static $iteratorProperties = null;
+
+    /**
+     * Overlay variables. Take precedence over anything from the current scope
+     */
+    protected array $overlay;
+
+    /**
+     * Flag for whether overlay should be preserved when pushing a new scope
+     */
+    protected bool $preserveOverlay = false;
+
+    /**
+     * Underlay variables. Concede precedence to overlay variables or anything from the current scope
+     */
+    protected array $underlay;
+
+    public function __construct(
+        ?ViewLayerData $item,
+        array $overlay = [],
+        array $underlay = [],
+        ?SSViewer_Scope $inheritedScope = null
+    ) {
         $this->item = $item;
 
         $this->itemIterator = ($inheritedScope) ? $inheritedScope->itemIterator : null;
         $this->itemIteratorTotal = ($inheritedScope) ? $inheritedScope->itemIteratorTotal : 0;
         $this->itemStack[] = [$this->item, $this->itemIterator, $this->itemIteratorTotal, null, null, 0];
-    }
 
-    /**
-     * Returns the current "active" item
-     *
-     * @return object
-     * @deprecated 5.4.0 use getCurrentItem() instead.
-     */
-    public function getItem()
```

Comment:
`getCurrentItem()` as a name for this method reflects the fact that you're not necessarily getting the `$item` property, you might be getting the current item in the scoped iterator.